core/tests/snapshots.rs

```diff
@@ -610,4 +625,244 @@ mod tests {
+                .unwrap();
+            }
+        }
+    }
+
+    fn make_full_snapshot_archive(
+        bank: &Bank,
+        snapshot_config: &SnapshotConfig,
+    ) -> snapshot_utils::Result<()> {
+        let slot = bank.slot();
+        info!("Making full snapshot archive from bank at slot: {}", slot);
+        let bank_snapshots = snapshot_utils::get_bank_snapshots(&snapshot_config.snapshot_path)
+            .into_iter()
+            .find(|elem| elem.slot == slot)
+            .ok_or_else(|| Error::new(ErrorKind::Other, "did not find snapshot with this path"))?;
+        let snapshot_package = snapshot_utils::package_full_snapshot(
+            bank,
+            &bank_snapshots,
+            &snapshot_config.snapshot_path,
+            bank.src.slot_deltas(&bank.src.roots()),
+            &snapshot_config.snapshot_package_output_path,
+            bank.get_snapshot_storages(),
+            snapshot_config.archive_format,
+            snapshot_config.snapshot_version,
+            None,
+        )?;
+        let snapshot_package = snapshot_utils::process_accounts_package_pre(
+            snapshot_package,
+            Some(bank.get_thread_pool()),
+            None,
+        );
+        snapshot_utils::archive_snapshot_package(
+            &snapshot_package,
+            snapshot_config.maximum_snapshots_to_retain,
+        )
+    }
+
+    fn make_incremental_snapshot_archive(
+        bank: &Bank,
+        incremental_snapshot_base_slot: Slot,
+        snapshot_config: &SnapshotConfig,
+    ) -> snapshot_utils::Result<()> {
+        let slot = bank.slot();
+        info!(
+            "Making incremental snapshot archive from bank at slot: {}, and base slot: {}",
+            slot, incremental_snapshot_base_slot,
+        );
+        let bank_snapshots = snapshot_utils::get_bank_snapshots(&snapshot_config.snapshot_path)
+            .into_iter()
+            .find(|elem| elem.slot == slot)
+            .ok_or_else(|| Error::new(ErrorKind::Other, "did not find snapshot with this path"))?;
+        let incremental_snapshot_package = snapshot_utils::package_incremental_snapshot(
+            bank,
+            incremental_snapshot_base_slot,
+            &bank_snapshots,
+            &snapshot_config.snapshot_path,
+            bank.src.slot_deltas(&bank.src.roots()),
+            &snapshot_config.snapshot_package_output_path,
+            bank.get_incremental_snapshot_storages(incremental_snapshot_base_slot),
+            snapshot_config.archive_format,
+            snapshot_config.snapshot_version,
+            None,
+        )?;
+        let incremental_snapshot_package = snapshot_utils::process_accounts_package_pre(
+            incremental_snapshot_package,
+            Some(bank.get_thread_pool()),
+            Some(incremental_snapshot_base_slot),
+        );
+        snapshot_utils::archive_snapshot_package(
+            &incremental_snapshot_package,
+            snapshot_config.maximum_snapshots_to_retain,
+        )
+    }
```

Comment:
hmm is there a reason this can't be encapsulated within a single call to the `bank_to_incremental_snapshot_archive()` function? Would prefer to call/test the top-level function instead of replicating that logic

Reply:
@carllin Yeah, I would prefer that too. I did it this way for a few reasons. Firstly, that's how the existing tests in this file are written w.r.t. full snapshots. Second, and more importantly, the `bank_to_xxx_snapshot_archive()` is _not_ the function that gets called during regular node operation. This test is sort of simulating the regular operation of going through AccountsBackgroundService and the calls it would make. And since this is an integration test, I like having that path tested too.

Some additional info, `bank_to_xxx_snapshot_archive()` is only called (1) when the node first starts up (during validator `new_banks_from_ledger()`, and (2) in ledger-tool when calling it with "create-snapshot". So the `bank_to_xxx_snapshot_archive()` function is not the commonly-used one for making snapshots (even though it's the nicer one to use).

More info: these functions getting called _are_ top-level functions (in the sense that they are `pub`), because of the Second reason above. The tests in snapshot_utils kind of test these functions, but more in an unit-test manor. The simulated AccountsBackgroundService is really the additional benefit here.

With that, are there still changes you'd like?

Reply:
Hmmm, from my understanding, `bank_to_xxx_snapshot_archive` reflects the path of the regular operation through `AccountsBackgroundService` (but I could be missing something).

The call stacks look rather similar:

```
make_incremental_snapshot_archive
get_bank_snapshots
package_incremental_snapshot
process_accounts_package_pre
archive_snapshot_package

bank_to_incremental_snapshot_archive
get_incremental_snapshot_storages
add_bank_snapshot
package_incremental_snapshot
process_accounts_package_pre
archive_snapshot_package
```
If these are close approximations of each other, I would prefer we coalesce this logic so that if we modify any of these intermediary steps in the snapshotting process, we only have to modify one function rather than multiple

Reply:
Regular node operation for the snapshots is as follows:

- `BankForks::set_root()` will call `AbsRequestSender::send_snapshot_request()` if the bank's slot is a multiple of the `accounts_hash_interval_slots`. This puts the (bank) snapshot request into the `SnapshotRequest` channel.
- `AccountsBackgroundService`'s main loop calls `AbsRequestHandler::handle_snapshot_request()`, which then calls `SnapshotRequestHandler::handle_snapshot_request()`
    - if there's a snapshot request in the channel, it grabs it, calls `Bank::clean_accounts()`, then calls `snapshot_utils::snapshot_bank()` (which creates a bank snapshot by serializing the bank to disk)
        - inside `snapshot_utils::snapshot_bank()`, it calls `add_bank_snapshot()` to do the bank serialization, and then it makes a snapshot package (`AccountsPackagePre`) and puts that in the channel for snapshot packages
- `AccountsHashVerifier`'s main loop reads from the snapshot packages channel, then
    - calls `snapshot_utils::process_accounts_package_pre()`
    - gets the hash from the processed package ^^ and checks it against the trusted validators
    - if the block height is a multiple of the `snapshot_interval_slots` then it puts this snapshot package into the `PendingSnapshotPackage` (`pub type PendingSnapshotPackage = Arc<Mutex<Option<AccountsPackage>>>`)
- Lastly, `SnapshotPackagerService`'s main loop takes the pending snapshot package (if there is one) and archives the snapshot package via `snapshot_utils::archive_snapshot_package()`

Since there are four separate threads handling the different parts of the snapshot process, it's not possible to just call `bank_to_xxx_snapshot()` during normal operation. So the `bank_to_xxx_snapshot()` functions are basically just wrappers around the other functions above, for the two times that making a snapshot archive can happen sequentially/blocking (1. at startup, 2. ledger-tool "create-snapshot").

I'm definitely in favor of coalescing the logic, I just am not sure how to do that here. I'm guessing I'm missing something though; I'd like to simplify this process if at all possible. Thank you in advance for the additional info!

Reply:
Carl and I talked. I've created a new issue to track this request: #18972

Reply:
OK, done!